source/retryable-reads/tests/README.rst

```diff
@@ -0,0 +1,269 @@
+
+    - ``arguments``: Optional, the names and values of arguments.
+
+    - ``result``: The return value from the operation, if any. This field may
+      be a scalar (e.g. in the case of a count), a single document, or an array
+      of documents in the case of a multi-document read. If the operation is
+      expected to return an error, the ``result`` is a single document that has
+      one or more of the following fields:
+
+      - ``error``: Optional. If ``true``, the test should expect an error or
+        exception. Implicitly true if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or
+        ``errorLabelsOmit``. If ``false``, the test should expect no error or
+        exception, even if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or ``errorLabelsOmit``.
+
+      - ``errorContains``: Optional. A substring of the expected error message.
+
+      - ``errorCodeName``: Optional. The expected "codeName" field in the server
+        error response.
+
+      - ``errorLabelsContain``: Optional. A list of error label strings that the
+        error is expected to have.
+
+      - ``errorLabelsOmit``: Optional. A list of error label strings that the
+        error is expected not to have.
+
+  - ``expectations``: Optional list of command-started events.
+
+MapReduce Tests
+===============
+
+Drivers MUST ensure that ``mapReduce`` is retried only when ``out: { inline:
+1}``.  The tests for ``mapReduce`` should test against the same conditions as
+described in ``aggregate.yml`` and ``aggregate-serverErrors.yml``, with the
+exception that ``mapReduce`` MUST not retry when the `out` stage is not
+``{inline: 1}``.
+
+A recommended data set would be three documents that contain values for x such
+that ``x:0``, ``x:1`` and ``x:2``. The map function should increment each of
+these values by one, and the reduce function should add all x's together.
+    
+GridFS Tests
+============
+
+Since the GridFS API is implemented using ``find`` commands, the `File
+Download`_ and `Generic Find on File Collection`_ parts of the GridFS API should
+be retryable under the same conditions that ``find`` is retryable. These
+functions should be tested against the same conditions described in ``find.yml``
+and ``find-ServerErrors.yml``. The tests' setup and expectations can be adapted
+from the "Download by Name when revision is 0" test at
+https://github.com/mongodb/specifications/blob/master/source/gridfs/tests/download_by_name.yml
+and the "Download when there is one chunk" test at
+https://github.com/mongodb/specifications/blob/master/source/gridfs/tests/download.yml.
+
+.. _File Download: https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#file-download-by-filename
+
+.. _Generic Find on File Collection:  https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#generic-find-on-files-collection
```

Comment:
On second thought, I dunno if it's worth even having this as a prose test. If drivers know that their GridFS implementation uses the same code path as `collection.find()`, there shouldn't be anything to test. Also, file downloads may or may not entail `getMore` commands, which makes it likely that only small file downloads (i.e. single batch of results) will survive a retryable error when reading. Drivers cannot reliably guarantee complete retryability for GridFS read operations and users may not even be aware of whether a file read entails `getMore` commands.

Reply:
If all drivers know that their GridFS implementation uses the same code path as `collection.find(),` I would be comfortable simply noting this in the test plan/spec as why there are no explicit tests for GridFS. However, if we were committed to testing the GridFS API for retryability, it would only be the initial find that would be on the files collection that would be retried (and we could add this note to the test plan).

Reply:
> We would only be retrying if the find on the files collection fails, I think.

Initial `find` on the chunks collection could also be retried. Anything that isn't a `getMore` is retryable.

In any event, GitHub says the diff for this thread is outdated so I'll resolve this. I don't feel strongly either way, but my preference is not to bother testing GridFS (we didn't do so for retryable writes, either).